x-pack/plugins/alerting/server/lib/transform_action_params.ts

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Handlebars from 'handlebars';
+import { isPlainObject } from 'lodash';
+import { AlertActionParams, State, Context } from '../types';
+
+export function transformActionParams(params: AlertActionParams, state: State, context: Context) {
+  const result: AlertActionParams = {};
+  for (const key of Object.keys(params)) {
+    const value = params[key];
+    if (isPlainObject(value)) {
+      result[key] = transformActionParams(value as AlertActionParams, state, context);
+      continue;
+    }
+    if (typeof value !== 'string') {
+      result[key] = value;
+      continue;
+    }
+    const template = Handlebars.compile(value);
+    result[key] = template({ context, state });
```

Comment:
Still don't think this may be as flexible as I originally thought, think of scenarios where array of items need to be passed to an action.

Reply:
Ya, this could get complicated.  I was wondering if we wanted to pass the "action params" which would include the uninterpreted templates, and the "context" to the action, and let the action side do the interpolation.  Just moves the complicated bits there, doesn't make the problem "easier".  However, if Actions will have some life outside of Alerts (some day), being able to do the interpolation there could be pretty useful.

Reply:
BTW, here's an example of a very rich "object" that in a perfect world you could construct for a Slack message: https://api.slack.com/incoming-webhooks#advanced_message_formatting ; I don't think we'll get there any time soon (if ever), but figure I'd provide an example.  Pager duty incident creation (presumably what the PD action will do), is similarly rich.

Reply:
Just writing down our offline discussion, we both agree the limitations of using templates will be reached very fast. We have decided to move forward with this for now and revisit this in a future PR.